### ISSUE TITLE
Android Application freezes up completely when using saveEventually()

### ISSUE THREAD

**Issue by tutorbear** [parse-community/Parse-SDK-Android#1007]
### Issue Description

saveEventually() is suppose to save objects using relations recursively just like saveInBackground does. The logic of my code is very simple and straight forward. I have a logged in user. I create a new ParseObject (Profile) and associate it with the user. Then when I call saveEventually The entire app freezes up. 

### Steps to reproduce

1) Signup and login.
2) Put this code on the onClick listener of a button in android.

      ` Parse user = ParseUser.getCurrentUser();`
       ` //A profile object`
       `profile = new ParseObject("StudentProfile");`
      `  profile.put("user", user);`
      `  profile.put("name", "John");`
       ` profile.setACL(new ParseACL(user));`
        `//Put profile object into user`
       ` user.put("profile",profile);`
        `user.saveEventually();`
### Expected Results

user saved along with new profile object while pointing towards it.

### Actual Outcome

App Freezes

### Environment Setup
  - parse-server version: 3.9.0
  - Hosted on Back4app

### Logs/Trace

Non , App freezes up and stops responding

**Comment by Jawnnypoo** [comment]
Can you pinpoint the exact line that causes the freeze, and can you see if `saveInBackground` causes the same freeze?

**Comment by ramiro-ciocca** [comment]
@jawnnypoo, I don't know if @tutorbear problem is related with the issues I have opened about LDS and saves.
I see that @tutorbear  is trying to save an object with a pointer to an usaved object, maybe that is his problem.
But, there is a real issue with LDS and saves.  If you LDS grows enough, and you try to save some complex object, with some pointers to other objects, then the app hangs.
I explained in #1006 and in #993 that what causes the hangs is a recursive method inside ParseObject that inefficiently traverse all the LDS with each save.

**Comment by tutorbear** [comment]
@Jawnnypoo 
Hello.
SaveInBackgound() works fine, its save eventually thats causing the issue. I have tried it with both LDS enabled and disabled , on both occasion the app freezes up. And I have found the exactly whats wrong.

`Parse user = ParseUser.getCurrentUser();`
`//A profile object`
`profile = new ParseObject("StudentProfile");`

`// THIS LINE is the curprit----------------------->
`profile.put("user", user);`
`//-------------------------------------->

`profile.put("name", "John");`
`profile.setACL(new ParseACL(user));
`//Put profile object into user
`user.put("profile",profile);
`user.saveEventually();



No error is givin in the log cat, only this is shown
`2020-02-26 12:10:13.832 9760-9773/com.example.tester I/.example.teste: Background concurrent copying GC freed 346064(19MB) AllocSpace objects, 0(0B) LOS objects, 49% free, 20MB/41MB, paused 338us total 191.114ms`
`2020-02-26 12:10:15.840 9760-9773/com.example.tester I/.example.teste: Background concurrent copying GC freed 372128(21MB) AllocSpace objects, 0(0B) LOS objects, 49% free, 21MB/43MB, `

@ramiro-ciocca  Like ramiro said complex pointer saving is causing the issue i guess ?

**Comment by tutorbear** [comment]
So no one is going to do anything about this issue ? Has the android skd community given up on fixing bugs  ?

**Comment by ramiro-ciocca** [comment]
Hello, like I said in previous comment, in your particular case, I think that trying to save a parse object with a pointer to an unsaved parse object can lead to problems.
Said that, we know that local data store has a lot of issues, and in certains conditions it collapses apps.
The main problem are the saves in all their variants, that by default try to save in cascade the root object and all its pointers, traversing all the local data store recursivelly, looking for dirty children, and hanging the app.
Like I said in previous posts, I have succesfully implemented a "cascade save false" variant of the sdk, that prevents the hangs. With this fix, you have to take care about saving the root object and every children you need to save as well. No cascade save, but no hangs.

**Comment by tutorbear** [comment]
Yes I too have found a work around , but its sad that we have do do work arounds while the community is still there.
Yes maybe it is I guess ? No one wants to take up a task this big. Since we have found work arounds , they chose to ignore it I guess. I just feel sorry about the fact that this will always remain buggy. Apart from this one problem I have found many others related to the LDS. Have a look at this post.
https://github.com/parse-community/Parse-SDK-iOS-OSX/issues/535
This issue was first posted in the year 2005 and still pending answers lol. Anyone and everyone has just gone through work arounds and thats it. Very sad in my opinion

**Comment by mtrezza** [comment]
> Too much to ask ? Yes maybe it is I guess ? No one wants to take up a task this big. Since we have found work arounds , they chose to ignore it I guess.

First of all, thank you for taking the time and bringing attention to the issue. Bug reporting and analysis are important first steps to fix an issue, and we are glad to have you in the community. As you already mentioned, this is a community effort. Do you want to take the next step and dig deeper into this to open a PR?

**Comment by Shvet** [comment]
Hello, I have same problem as above `saveEventually()` keep code in loop and garbage collector is running with it which is giving me application isn't responding error.

**Comment by mtrezza** [comment]
It would be great if someone could open a PR with a failing test to replicate this.

**Comment by Shvet** [comment]
What do i have to add for PR? I can create simple android code for this but on sever side i can not share info as it would not be mine. @mtrezza

**Comment by L3K0V** [comment]
@Shvet which version of the SDK are you using, can you update the issue description?

**Comment by Shvet** [comment]
@mtrezza Give me time i will create simple login/register and update project with user object and upload it.

**Comment by mtrezza** [comment]
> What do i have to add for PR? I can create simple android code for this but on sever side i can not share info as it would not be mine.

Server code should not be needed; see the tests in the existing test directory to see how to write a new test.

**Comment by Shvet** [comment]
When i tried testing it, it no longer freezes screen.

**Comment by mtrezza** [comment]
Could it be related to your custom code rather than the Parse SDK? You could try to reduce your project down to the least code needed to reproduce the issue. Cut out all UI and specially thread related code until you cannot reproduce the issue anymore. That should reveal whether the issue lies really with the Parse SDK.
